packages/react-utilities/src/compose/types.ts

```diff
@@ -16,17 +16,64 @@ export type ShorthandProps<Props extends DefaultObjectShorthandProps> =
   | undefined // default render (or null render if no default provided)
   | Props;
 
-export type DefaultObjectShorthandProps = ObjectShorthandProps<{}, unknown, keyof JSX.IntrinsicElements>;
-
-export type ObjectShorthandProps<
-  Props extends { children?: React.ReactNode } = {},
-  Ref = unknown,
-  As extends keyof JSX.IntrinsicElements = never
-> = Props &
-  React.RefAttributes<Ref> & {
-    as?: As;
-    children?: Props['children'] | ShorthandRenderFunction<Props>;
-  };
+/**
+ * Matches any shorthand props type.
+ *
+ * This should ONLY be used in type templates as in `extends DefaultObjectShorthandProps`;
+ * it shouldn't be used as the type of a slot.
+ */
+export type DefaultObjectShorthandProps = ObjectShorthandProps<{
+  children?: React.ReactNode;
+  as?: keyof JSX.IntrinsicElements;
+}>;
+
+/**
+ * Defines the slot props for a slot that supports a Component type.
+ *
+ * For intrinsic elements like 'div', use {@link ElementShorthandProps} instead.
+ */
+export type ObjectShorthandProps<Props extends { children?: React.ReactNode } = {}> = Props & {
+  children?: Props['children'] | ShorthandRenderFunction<Props>;
+};
+
+/**
+ * Define the slot arguments for a slot that supports one or more intrinsic element types, such as 'div'.
+ * For slots that support custom components, use {@link ObjectShorthandProps} instead.
+ *
+ * The first param is the slot's default type if no `as` prop is specified.
+ * The second param is an optional union of alternative types that can be specified for the `as` prop.
+ *
+ * ```
+ * ElementShorthandProps<'div'> // Slot is always a <div>
+ * ElementShorthandProps<'a', 'button'> // Defaults to <a>, but allows { as: 'button' } to be a <button>
+ * ElementShorthandProps<'label', 'span' | 'div'>; // Defaults to <label>, but allows { as: 'span' } or { as: 'div' }
+ * ```
+ */
+export type ElementShorthandProps<
```

Comment:
I'm open to suggestions about the name of `ElementShorthandProps`. Possible alternatives include:
* `ObjectShorthandPropsAs`
* `ObjectShorthandElementProps`
* `ObjectShorthandIntrinsicElement`

Reply:
`IntrinsicShorthandProps` or `IntrinsicElementShorthandProps` seems better to me because of the `JSX.IntrinsicElements` thing

Reply:
I'll rename it to `IntrinsicShorthandProps`  for this PR. 

However, I was just thinking about the name `ObjectShorthandProps` and it really doesn't make sense: it is the _non-shorthand_ version of the props, so it shouldn't include "Shorthand" in the name. Same for `IntrinsicShorthandProps`: it is also non-shorthand.

I'll follow up with another PR to keep it separate (and because it might generate some discussion), but I'm thinking of renaming:
* ObjectShorthandProps -> SlotProps
* IntrinsicShorthandProps -> SlotPropsAs
* ShorthandProps -> ShorthandSlotProps